hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -0,0 +1,533 @@
+
+    val modMap = clsBmsMap
+      .map:
+        case b1 -> b2 => b1.asMod match
+          case Some(value) => 
+            val newSym = ModuleSymbol(value.tree, Tree.Ident(b2.nme))
+            newSym.defn = value.defn
+            S(value -> newSym)
+          case None => None
+      .collect:
+        case Some(x) => x
+      .toMap
+    
+    val newBlk = defns.foldLeft(blk)((acc, defn) => Define(defn, acc))
+
+    given SymbolSubst with
+      override def mapBlockMemberSym(b: BlockMemberSymbol) = bmsMap.get(b) match
+        case None => b.asCls match
+          case None => b
+          case Some(cls) => 
+            clsMap.get(cls) match
+              case None => b
+              case Some(sym) => 
+                BlockMemberSymbol(sym.nme, b.trees) // TODO: properly map trees
+        case Some(value) => value
+      override def mapClsSym(s: ClassSymbol): ClassSymbol = clsMap.get(s).getOrElse(s)
+      override def mapModuleSym(s: ModuleSymbol): ModuleSymbol = modMap.get(s).getOrElse(s)
+      override def mapTermSym(s: TermSymbol): TermSymbol = TermSymbol(s.k, s.owner.map(_.subst), s.id)
+
+    newBlk.mapSyms
+  
+  private def translateFun(f: FunDefn): FunDefn =
+    FunDefn(f.sym, f.params, translateBlock(f.body, functionHandlerCtx))
+  
+  private def translateCls(cls: ClsLikeDefn): ClsLikeDefn =
+    cls.copy(methods = cls.methods.map(translateFun), ctor = translateBlock(cls.ctor, functionHandlerCtx))
+  
+  // Handle block becomes a FunDefn and CallPlaceholder
+  private def translateHandleBlock(h: HandleBlock): Block =
+    val sym = BlockMemberSymbol(s"handleBlock$$${freshId()}", Nil)
+    val lbl = freshTmp("handlerBody")
+    val lblLoop = freshTmp("handlerLoop")
+    val tmp = freshTmp("retCont")
+    def prepareBody(b: Block): Block =
+      def go(b: Block): Block =
+        b.map(go) match
+        case Return(res, implct) =>
+          // In the case of res is effectful, it will be handled in translateBlock
+          Assign(tmp, res, Return(Instantiate(retClsPath, tmp.asPath :: Nil), implct))
+        case HandleBlockReturn(res) =>
+          Return(res, false)
+        case b => b
+      go(b)
+    val handlerBody = translateBlock(prepareBody(h.body), HandlerCtx(false, false, state => blockBuilder
+      .assignFieldN(state.res.tail, nextIdent, Instantiate(state.cls, Value.Lit(Tree.IntLit(state.uid)) :: Nil))
+      // .assignFieldN(state.res, tailIdent, state.res.tail.next)
+      .ret(SimpleCall(handleBlockImplPath, state.res :: h.lhs.asPath :: Nil))))
+    
+    val handlers = h.handlers.map: handler =>
+      val lam = Value.Lam(PlainParamList(Param(FldFlags.empty, handler.resumeSym, N) :: Nil), translateBlock(handler.body, functionHandlerCtx))
+      val tmp = freshTmp()
+      FunDefn(handler.sym, handler.params, Return(SimpleCall(mkEffectPath, h.lhs.asPath :: lam :: Nil), false))
+    
+    // TODO: it seems that our current syntax didn't know how to call super, calling it with empty param list now
```

Comment:
Can you elaborate? Maybe open a GitHub issue with a minimal example.

Reply:
It is the syntax `handle h = Foo(123)` as right now it is just calling it without any parameters. Should this syntax be also implemented inside this PR?

Reply:
Not necessarily. Better do it in a follow-up.

Reply:
But the comment could be clarified.